plugin/vscode/performance-writer.js

```diff
@@ -0,0 +1,93 @@
+// See end of file for extended copyright information.
+
+const fs = require('fs');
+const path = require("path");
+const process = require("process");
+
+class PerformanceWriter {
+    
+    constructor (extensionContext) {
+        this._dataAttributes = {
+            columns:'timestamp,file-ref,file-name-or-duration,event-id'
+        };
+        this._tmpLogLines = []
+        this._fileName = null;
+        this._fileIndex = null;
+        this._extensionContext = extensionContext;
+        fs.mkdirSync(this._extensionContext.logUri.fsPath)
+        this._performanceTxtPath = path.join(this._extensionContext.logUri.fsPath, "qljsPerformanceTrace.txt");
+        this._currentSessionMap = this._extensionContext.globalState.get('qljsFileMap');
+        this._writableStream = fs.createWriteStream(this._performanceTxtPath, {encoding: 'utf8', flags: 'a'})
+    }
+
+    createHeader (){
+        this.writeToTmpArray(null,"header");
+    }
+    
+    fileNameOptimizer (){
+        this._fileName = this._fileName.split(path.sep);
+        this._fileName = this._fileName[this._fileName.length - 1];
+        if (!(this._currentSessionMap)) {
+            this._currentSessionMap = {};
+            this._currentSessionMap['*latestIndex'] = 1;
+            this._currentSessionMap[this._fileName] = 1;
+            this._fileIndex = 1;
+            
+            this.writeToTmpArray(null, "newFile");
+            this._extensionContext.globalState.update('qljsFileMap', this._currentSessionMap);
+            return 'newFile'
+        } else {
+            if (Object.prototype.hasOwnProperty.call(this._currentSessionMap, this._fileName)) {
+                this._fileIndex = this._currentSessionMap[this._fileName];
+            } else {
+                this._currentSessionMap['*latestIndex'] = this._currentSessionMap['*latestIndex'] + 1;
+                this._currentSessionMap[this._fileName] = this._currentSessionMap['*latestIndex'];
+                this._fileIndex = this._currentSessionMap['*latestIndex'];
+                this._extensionContext.globalState.update('qljsFileMap', this._currentSessionMap);
+                this.writeToTmpArray(null, "newFile", this._currentSessionMap['*latestIndex']);
+            }
+        }
+    }
+    
+    writeToTmpArray(timeTaken=null, mode=null) {
+        if (mode === 'header') {
+            this._tmpLogLines.push(this._dataAttributes.columns);
+        } else if (mode === "newFile") {
+            this._tmpLogLines.push('\n' + [Date.now(), this._fileIndex, this._fileName, 'n'].join(','));
+        } else {
+            this._tmpLogLines.push('\n' + [Date.now(), this._fileIndex, timeTaken, ''].join(','));
+        }
+    }
+
+    writeToFile() {
+        let tracingLine = null
+        if (this._tmpLogLines.length !== 0) {
+            for (tracingLine of this._tmpLogLines) {
+                this._writableStream.write(tracingLine)
+            }
+            this._tmpLogLines = []
```

Comment:
Performance: I *think* the following is better, but I'm not sure:
```suggestion
            this._tmpLogLines.length = 0;
```

Reply:
[Addressed in latest commit]

This line is no longer relevant in latest commit